web/packages/teleterm/src/mainProcess/clusterLifecycleManager/clusterLifecycleManager.ts

```diff
@@ -0,0 +1,227 @@
+import { RootClusterUri } from 'teleterm/ui/uri';
+
+import { ClusterStore } from '../clusterStore';
+import { ProfileChangeSet } from '../profileWatcher';
+
+export interface ClusterLifecycleEvent {
+  uri: RootClusterUri;
+  op: 'did-add-cluster' | 'will-logout' | 'will-logout-and-remove';
+}
+
+export interface ProfileWatcherError {
+  error: unknown;
+  reason: 'processing-error' | 'exited';
+}
+
+export class ClusterLifecycleManager {
+  private readonly logger = new Logger('ClusterLifecycleManager');
+  private rendererEventHandler:
+    | AwaitableSender<ClusterLifecycleEvent>
+    | undefined;
+  private watcherStarted = false;
+
+  constructor(
+    private readonly clusterStore: ClusterStore,
+    private readonly getTshdClient: () => Promise<TshdClient>,
+    private readonly appUpdater: AppUpdater,
+    private readonly windowsManager: Pick<WindowsManager, 'getWindow'>,
+    private readonly profileWatcher: AsyncIterable<ProfileChangeSet>
+  ) {}
+
+  setRendererEventHandler(
+    handler: AwaitableSender<ClusterLifecycleEvent>
+  ): void {
+    if (this.rendererEventHandler) {
+      this.logger.error(
+        'Only one renderer lifecycle event handler can be registered at a time'
+      );
+      return;
+    }
+
+    this.logger.info('Renderer lifecycle event handler registered');
+    this.rendererEventHandler = handler;
+    this.rendererEventHandler.whenDisposed().then(() => {
+      this.logger.info('Renderer lifecycle event handler unregistered');
+      this.rendererEventHandler = undefined;
+    });
+  }
+
+  async addCluster(proxyAddress: string): Promise<Cluster> {
+    const cluster = await this.clusterStore.add(proxyAddress);
+    await this.rendererEventHandler.send({
+      op: 'did-add-cluster',
+      uri: cluster.uri,
+    });
+    return cluster;
+  }
+
+  async logoutAndRemoveCluster(uri: RootClusterUri): Promise<void> {
+    this.onBeforeLogout(uri);
+    // This method is currently only called from the renderer.
+    // Because of that, if this.rendererEventHandler.send() throws, we return the error.
+    // This is different from error handling in the profile watcher,
+    // where renderer errors don't prevent updating the cluster store.
+    await this.rendererEventHandler.send({ op: 'will-logout-and-remove', uri });
+    await this.clusterStore.logoutAndRemove(uri);
+  }
+
+  async syncRootClustersAndStartProfileWatcher(): Promise<void> {
+    await this.clusterStore.syncRootClusters();
+    if (!this.watcherStarted) {
+      this.watcherStarted = true;
+      void this.watchProfileChanges();
+    }
+  }
+
+  private onBeforeLogout(uri: RootClusterUri): void {
+    // This function checks for updates, do not wait for it.
+    this.appUpdater.maybeRemoveManagingCluster(uri).catch(error => {
+      this.logger.error('Failed to remove managing cluster', error);
+    });
+  }
+
+  /**
+   * If the cluster is connected, try to sync it to get details.
+   * Otherwise, only update profile properties.
```

Comment:
What's the reason for only updating profile properties if the cluster is not connected? Wouldn't the renderer already have those up to date from the previous sync?

Edit: I guess I understand it a little bit better now that I spotted that the argument is of type `Cluster`. So it likely is the profile from disk read by the profile watcher. But when I was first reading this method, this flow of information wasn't quite clear for me. Maybe just updating the comment to something like this would help:

> If the cluster is connected, try to sync it to get the full profile with details.
> Otherwise, update the cluster with the profile read from disk.

Reply:
I updated the comment, I think it's more clear now.